### DISCUSSION TITLE
[WASM][Debug][Windows 11] App no longer can be debugged

### DISCUSSION THREAD

**Discussion by DierkDroth** [unoplatform/uno#7999] [Q&A] [answered]
I migrated my WASM solution from VisualStudio 2022 on Windows 10 to a new laptop with VisualStudio 2022 on Windows 11 and found that I no longer could start the app in DEBUG but saw tons of errors (below) in VS output window.

This made me tinker with launchSettings.json and compare it with the version of the sample solution which I was using [here](https://github.com/unoplatform/uno/issues/7974) and which - to my big surprise - still did start on Windows 11 (however, I saw tons of errors in browser console but that's a different story likely related to the wrong default settings in 'launchSettings.json/iisSettings')

I finally found that adding
```
"inspectUri": "{wsProtocol}://{url.hostname}:{url.port}/_framework/debug/ws-proxy?browser={browserInspectUri}",
```
... as advised [here ](https://platform.uno/docs/articles/debugging-wasm.html?q=inspectUri) was the problem (interestingly these lines weren't added by default by the [generated working solution above](https://github.com/unoplatform/uno/issues/7974), again another issue which should be looked into). Would I remove/add those lines to any solution I tested then the WASM head would work/not work.

Removing those lines got the WASM app started, but now I no longer could e.g. set breakpoints -> bad.

@jeromelaban any idea what needed to be done to get me going with proper WASM debugging on Windows 11?

```
System.InvalidOperationException: No output has been recevied from the application.
   at Uno.Wasm.Bootstrap.Cli.DebugProxyLauncher.LaunchAndGetUrl(IServiceProvider serviceProvider, IConfiguration configuration, String devToolsHost, Uri browserUrl) in D:\a\1\s\src\Uno.Wasm.Bootstrap.Cli\DebuggingProxy\DebugProxyLauncher.cs:line 126
   at Uno.Wasm.Bootstrap.Cli.DebugProxyLauncher.EnsureLaunchedAndGetUrl(IServiceProvider serviceProvider, IConfiguration configuration, String devToolsHost, Uri browserUrl) in D:\a\1\s\src\Uno.Wasm.Bootstrap.Cli\DebuggingProxy\DebugProxyLauncher.cs:line 60
   at Uno.Wasm.Bootstrap.Cli.DebuggingProxy.DebuggerHostExtensions.<>c__DisplayClass0_0.<<UseWebAssemblyDebugging>b__1>d.MoveNext() in D:\a\1\s\src\Uno.Wasm.Bootstrap.Cli\DebuggingProxy\DebugHostExtensions.cs:line 66
--- End of stack trace from previous location ---
   at Microsoft.AspNetCore.Builder.Extensions.MapMiddleware.InvokeCore(HttpContext context, String matchedPath, String remainingPath)
   at Uno.Wasm.Bootstrap.Cli.Server.Startup.<>c.<<Configure>b__3_0>d.MoveNext() in D:\a\1\s\src\Uno.Wasm.Bootstrap.Cli\Server\Startup.cs:line 66
--- End of stack trace from previous location ---
   at Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddleware.Invoke(HttpContext context)
Microsoft.AspNetCore.Hosting.Diagnostics: Information: Request finished HTTP/1.1 GET https://localhost/_framework/debug/ws-proxy?browser=ws%3A%2F%2F127.0.0.1%3A56450%2Fdevtools%2Fbrowser%2F15b7b7cb-1815-4415-8bde-22aa58db960f - - - 500 - text/plain;+charset=utf-8 12.4544ms
Microsoft.AspNetCore.Hosting.Diagnostics: Information: Request starting HTTP/1.1 GET https://localhost/_framework/debug/ws-proxy?browser=ws%3A%2F%2F127.0.0.1%3A56450%2Fdevtools%2Fbrowser%2F15b7b7cb-1815-4415-8bde-22aa58db960f - -
The program '' has exited with code 4294967295 (0xffffffff).
Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddleware: Error: An unhandled exception has occurred while executing the request.
```

**Comment by jeromelaban** [comment]
Windows 11 is supported properly. This error message is likely related to some environment issue that prevents the debugger proxy from starting properly. This process is built from the .NET SDK and is likely to fail for non-Uno projects as well.

You can try starting the the debugger proxy manually from a path that looks like this one:
```
dotnet "%temp%\dotnet-runtime-wasm-windows-XXXXXXXXX-Release\dbg-proxy\net6.0\BrowserDebugHost.dll"
```

and see if there are unusual error messages showing up.

Could you also indicate which version of VS 2022 is installed on your machine ?

**Reply by DierkDroth** [reply]
Thanks for follow up @jeromelaban 

This is VS 2022 17.0.5

There are 2 such directories with identical names on both old and new laptop (screenshot below is from new laptop = not working). There is a diffence between old and new latop: on old laptop 'Content root path = C:\WINDOWS\system32' (not sure if that matters). Other than that output looks identical on old and new laptop, apart from port numbers which keep changing on every incarnation


![image](https://user-images.githubusercontent.com/26179242/152643262-29d57900-a2c3-4a11-9064-bd4f054a938f.png)

**Comment by DierkDroth** [comment]
@jeromelaban I'm still stranded. What else could be done to get me going with UNO WASM again?

Also: not sure if this helps, but the very first exception I do see in my VS output window is this
`Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddleware: Error: An unhandled exception has occurred while executing the request.`

**Reply by jeromelaban** [reply]
There's nothing additional to do aside from investigating the issue. It could be an environment issue, a browser issue, a Visual Studio issue or a .NET issue, but without a repro, it's difficult to troubleshoot.

**Comment by DierkDroth** [comment]
@jeromelaban the repo which I provided [here ](https://github.com/unoplatform/uno/issues/7974)(also please see above) exhibits the exact same issue. 

Also: the issue could be easily reproduced by any random UNo WASM project following directions [here](https://platform.uno/docs/articles/get-started-winui3.html).

Note: you needed to add the offending lines as per above to launchSettings.json
`"inspectUri": "{wsProtocol}://{url.hostname}:{url.port}/_framework/debug/ws-proxy?browser={browserInspectUri}",`

**Reply by DierkDroth** [reply]
Also:
- I have the exact same Chrome version on both machines (Version 98.0.4758.82 (Official Build) (64-bit))
- I checked that VS 2022 -> Extensions are the same on both machines
- I have Administrator entitlements on both machines
- I just updated to latest VS 2022 17.0.6 on both machines
- I exported the VS configuration from old machine and imported it on new machine (using VS installer -> More)
- Lastly I went through a full VS repair cycle and tried again before and after importing my custom configuration (VS installer -> More)

Whatever I tried: **there is no way to get UNO WASM debugging working on Windows 11**
The browser address bar just does not change from 'about:blank' to 'localhost:<port>'. After as while VS gives up and throws errors below.

Not sure what and how I could check this (referring to your reply)
- .NET versions are the same
- environments are the same

Please advise. Thanks

<img width="1280" alt="image" src="https://user-images.githubusercontent.com/26179242/153128266-66d2ddc0-41e6-418e-9181-8f55314ed9ce.png">

**Reply by jeromelaban** [reply]
Thanks for the report. When you answer, make sure to stay in the same thread above, so it'll mark an answer once.

Many smaller fixes were added in .NET 6.0 upstream, those will be integrated in https://github.com/unoplatform/Uno.Wasm.Bootstrap/pull/474, we'll see if this helps.

**Reply by DierkDroth** [reply]
Thanks @jeromelaban 

To clarify: you still can't reproduce the issue on Windows 11 by steps I provided?

**Reply by jeromelaban** [reply]
I still can't. The nuget package is now published under 3.3.0-dev.2. Can you try and see if this helps ?

**Reply by DierkDroth** [reply]
Sorry @jeromelaban, I tried latest Uno (below) and still am running into the same problem

![image](https://user-images.githubusercontent.com/26179242/153338637-3921ce65-8e96-4508-b804-bfdeef379ca7.png)

**Reply by jeromelaban** [reply]
Thanks. Can you also try with a Visual Studio preview? Otherwise, the solution you provided seems to be running net5 and has a global.json file. Try removing the global.json file and update to net6.0.

**Reply by DierkDroth** [reply]
@jeromelaban correct, the repo linked above runs on .NET 5 - since this is what the UNo templates provide (I did not want to change that to not introduce variables to the case). However, my custom project runs on .NET 6 - and exhibits the same faulty behavior.

- I removed globals.json from the repo above -> problem persists
- I don't have a globals.json file on my custom project

**Reply by jeromelaban** [reply]
Thanks. Then other than a preview version, there's nothing else we can do at this point without reproducing.

**Reply by DierkDroth** [reply]
- What would I buy into if I installed a VS preview version? What other pieces of UNO would then fall apart?
- I don't even know how to install a VS preview. How would I do that? Is there a roll back if things go south?
- What happens on the final VS release? Could I just install it on top or would I have to go through a full uninstall/re-install cycle?

Also: 
- what is it that UNO is trying to connect to as it fails? What could be done to understand why the 'thing which it's talking to' would not behave as expected?
- could the failing part be simulated in some shape or form in an attempt to analyze what's going wrong?
- is the difference in 'Content root path' between old and new machine (which I outlined above) of any relevance? If so, how could I 'make that equal'?

**Reply by DierkDroth** [reply]
When trying to understand the UNO sources I see this in DebugProxyLauncher.cs:

```
	var processStartInfo = new ProcessStartInfo
	{
		FileName = muxerPath,
		Arguments = $"exec \"{executablePath}\" --OwnerPid {ownerPid} --DevToolsUrl {devToolsHost}",
		UseShellExecute = false,
		RedirectStandardOutput = true,
	};
```

How about amending sources to trace out or log the relevant variables so I could try to reproduce the case in my custom code in order to understand why the ASP.NET process eithes does not start or does not send any data?

Another idea: I don't know it it even would be possible that I debugged this part of the UNO WASM  bootstrapper directly?!?

**Reply by jeromelaban** [reply]
The parts that need to be debugged for this specific issue are directly from the .NET runtime and not the uno bootstrapper. Those communicate directly with Visual Studio which we cannot debug at all as they are private.

Here's the visual studio preview link: https://visualstudio.microsoft.com/vs/preview. You can use it side by side with stable builds of VS. A new release of VS 2022 should be released soon, so you cam wait for an update there.

Also, if you don't do anything specific to WebAssembly, debugging using the windows or GTK/WPF targets of your app is far more useful in all cases as the debugger has a lot more features.

**Comment by DierkDroth** [comment]
Thanks @jeromelaban 

> The parts that need to be debugged for this specific issue are directly from the .NET runtime and not the uno bootstrapper. Those communicate directly with Visual Studio which we cannot debug at all as they are private.

Hmm ... isn't the code which I provided above from UNO - and isn't it trying to start an ASP.NET process which apparently triggers .OutputDataReceived but 'string.IsNullOrEmpty(eventArgs.Data)' and hence the exception is thrown in DebugProxyLauncher.cs 188?

Believe me, I'm only debugging WASM as it's absolutely unavoidable (since WASM debugging still is close to a nightmare). However there are many scenarios where I needed to understand why stuff is working in WinUI, but not in WASM...

**Reply by jeromelaban** [reply]
Yes, the code provided is from Uno, but the failing process which seems to be not responding and that process is the one that is coming from the .NET runtime. You can also try to debug a WebAssembly Blazor application and see if it works, as it uses the same debugger binaries.

**Reply by DierkDroth** [reply]
Yep, that's why I suggested tracing out the parameters so I could replicate the scenario by some custom code - or even debug the very code to inspect the actual parameter values.

**Reply by DierkDroth** [reply]
@jeromelaban I just tried [the Blazor tutorial.](https://dotnet.microsoft.com/en-us/learn/aspnet/blazor-tutorial/run) That worked no problem.

Next I tried

`dotnet new blazorwasm -o blazor-web`

... and ran the project from VS 2022. That worked no problem as well.

There is a difference to the UNO setup though as there is an additional CMD window coming up (below).

Is there any other Blazor app or setup which I should try in order to isolate the problem?


![image](https://user-images.githubusercontent.com/26179242/153452247-987e9538-772d-4645-9d1d-1f9bc8537e54.png)

**Reply by DierkDroth** [reply]
@jeromelaban I just tried turning off my Windows Firewall -> make no difference, issue persists

So everything we have at this point - since it's not a firewall issue and blazor works no problem - is that UNO is trying to start the MS part in a way that would not work on my new laptop. 

I see no other way than inspecting/debugging the UNO code/logic to understand why it's failing. Since you can't reproduce either you had to add tracing/logging info to the UNO pieces (where I then could try to replicate the failing scenario in some custom code, to better understand why things fall aprt) or we had to debug the case on my machine...

**Reply by DierkDroth** [reply]
@jeromelaban what would be next steps in order to restore the debugging feature?

How about amending UNO sources to trace out or log the relevant variables on call below so I could try to reproduce the case in my custom code in order to understand why the ASP.NET process eithes does not start or does not send any data?

```var processStartInfo = new ProcessStartInfo
	{
		FileName = muxerPath,
		Arguments = $"exec \"{executablePath}\" --OwnerPid {ownerPid} --DevToolsUrl {devToolsHost}",
		UseShellExecute = false,
		RedirectStandardOutput = true,
	};
```

**Comment by DierkDroth** [comment]
I just came across something which we should look into. 

When trying to download and unzip the Uno.Wasm.Bootstrap sources from GitHub I got error below. I could have sworn I have enabled long path support before. Nevertheless I did it again as per [here ](https://platform.uno/docs/articles/get-started-wizard.html)by 

`reg ADD HKEY_LOCAL_MACHINE\SYSTEM\CurrentControlSet\Control\FileSystem /v LongPathsEnabled /t REG_DWORD /d 1`

... and rebooted.

But I'm still getting error below when trying to unzip the Uno.Wasm.Bootstrap sources?!?

**Is there a problem with long path support on Windows 11?!?**


![image](https://user-images.githubusercontent.com/26179242/153428259-5c51336c-61d1-4135-a602-3f21fd9ee0c5.png)

**Comment by DierkDroth** [comment]
@jeromelaban I just ran 
`uno-check`

...and

`uno-check --target wasm`

... and do see this (below). What does that mean and what should I be looking into?

![image](https://user-images.githubusercontent.com/26179242/153650016-437d65f7-1914-4702-b5d4-62bd0c5e53a3.png)

When running

`uno-check --verbose --target wasm`

... I do see those strange issues

![image](https://user-images.githubusercontent.com/26179242/153650720-f8c3bd8b-9324-4e84-8d03-e21e80510d67.png)

![image](https://user-images.githubusercontent.com/26179242/153650765-61b53b8e-80a8-4a82-8821-7d0ca7d17061.png)

**Comment by DierkDroth** [comment, marked as answer]
@jeromelaban (I purposely started a new thread)

- I just installed latest VS 2022 17.1.0 -> issue persists on my custom project and repo project above
- to double check I'm not missing out on something I installed the latest UNO DEV pieces -> issue persists on my custom project and repo project above

Next I ran more tests - unfortunately with very inconsistent results:
- I created a new test solution A as per [here ](url)and definitely was able to set a breakpoint in App.xaml.cs and **got it working once**
- Then I was working on my already existing repo project from above to reproduce the success ... but failed, came back to the solution A **without changing anything and the breakpoint no longer worked**
- Scratching my head I again created a new solution C as per [here](https://platform.uno/docs/articles/get-started-winui3.html) **did not touch anything and breakpoints weren't working again**
- On solution C I see tons of these errors in the browser console (below)

What needed to be done to get UNO WASM debuging working?

![image](https://user-images.githubusercontent.com/26179242/154487579-7bf4da87-3bdf-4cf7-acdb-54e84f59d27e.png)

**Reply by jeromelaban** [reply]
I'm still not able to reproduce your issue, but let's try with additional changes related to timeouts: https://github.com/unoplatform/Uno.Wasm.Bootstrap/pull/481. 

The errors that you are seeing are expected in an https context and are unrelated to debugging. You can disable https in the project properties.

**Reply by DierkDroth** [reply]
Thanks @jeromelaban I'll try that as the build is available.

Also: appears this line in launchSettings.json no longer is needed?
`"inspectUri": "{wsProtocol}://{url.hostname}:{url.port}/_framework/debug/ws-proxy?browser={browserInspectUri}",`

Is that correct?

**Reply by jeromelaban** [reply]
The `inspectUri` is required for visual studio to determine where to connect. Blazor automatically adds it, uno does not at this point.

**Reply by DierkDroth** [reply]
Understand, thanks. Will make sure it's added.

Curious: why is it no added by the default template created as per [here](https://platform.uno/docs/articles/get-started-winui3.html)?

**Reply by DierkDroth** [reply]
@jeromelaban unfortuantely I can't test your code change since there is a [new bug introduced](https://github.com/unoplatform/uno/issues/8131)

**Reply by jeromelaban** [reply]
Try with a stable build of uno first.

**Reply by DierkDroth** [reply]
@jeromelaban OK, I tried latest UNO stable packages -> issue persists. Please see attached

[Debugger2.zip](https://github.com/unoplatform/uno/files/8097548/Debugger2.zip)

**Reply by jeromelaban** [reply]
Which issue persists?

**Reply by DierkDroth** [reply]
@jeromelaban the original issue which this whole discussion is about: the WASM app does not start as soon as launchSettings.json holds the 'inspectUri' line.

Please let me know if I failed to make clear what the whole discussion is about. Thanks

**Reply by jeromelaban** [reply]
Thanks. Your sample above does not have the latest bootstrapper installed.

The debugger still works properly:

![image](https://user-images.githubusercontent.com/5839577/154729414-3bfda74e-df23-44b3-8a88-25836186ffff.png)

Where are you placing your breakpoints?

**Reply by DierkDroth** [reply]
@jeromelaban to clarify: you asked me to only use stable packages first (and not anything in pre-release mode), since - as per [my issue report](https://github.com/unoplatform/uno/issues/8131) - pre-releases don't work on the default UNO template.

That's what I did - please correct me if I'm wrong.
After doing that, I (still) see the original issue.

If that's not what you wanted me to do, then please be specific on steps you wanted me to execute from here on. Thanks

**Reply by jeromelaban** [reply]
I said above that the issue might be fixed by https://github.com/unoplatform/Uno.Wasm.Bootstrap/pull/481, so experimental build of Uno.Wasm.Bootstrap, that you need to use stable builds of .NET (no .NET 7 preview 1 as mentioned in https://github.com/unoplatform/uno/issues/8131), and any release of Uno.UI or Uno.WinUI above 4.0.x.

To ask again, where are you placing your break points ?

**Reply by DierkDroth** [reply]
@jeromelaban I updated the repo to latest Uno.Wam.Bootstrap (below + attached) and still experience the issue which is (again, since I feel we're not on the same page yet)
- I get tons of errors (below) in VS output
- the WASM app does not start at all
- browser address line stays in 'about:blank' and does not switch to 'localhost...'
- it make no sense to place any breakpoint, since the app does not start in the first place

Please let me know if it's still not clear what the problem is. Thanks

[Debugger2.zip](https://github.com/unoplatform/uno/files/8101570/Debugger2.zip)

![image](https://user-images.githubusercontent.com/26179242/154789614-ee3cdb3a-703d-4c25-adaf-641372ccd864.png)

![image](https://user-images.githubusercontent.com/26179242/154789501-8880ec09-2f92-406b-aecc-80446b082dfd.png)

**Reply by DierkDroth** [reply]
@jeromelaban after closing [this ](https://github.com/unoplatform/uno/issues/8131)and installing latest UNO NuGet packages (below) the issue still is there (please let me know if it's not clear what the issue is)

![image](https://user-images.githubusercontent.com/26179242/154811989-eb2b2885-0f1f-42e7-a6a3-7eb852ea8e96.png)

**Reply by jeromelaban** [reply]
Thanks for all the updates. I don't have any new information to share on the topic, as we're still not able to reproduce.

**Reply by jeromelaban** [reply]
To be able to determine if there's another runtime-related issue, can you paste the results of the `dotnet --list-sdks` command, as well as `dotnet --version` when started from the wasm project folder.

**Reply by DierkDroth** [reply]
Thanks for follow up @jeromelaban 

I suggested above that UNO would add traces to DebugProxyLauncher.cs:
```
	var processStartInfo = new ProcessStartInfo
	{
		FileName = muxerPath,
		Arguments = $"exec \"{executablePath}\" --OwnerPid {ownerPid} --DevToolsUrl {devToolsHost}",
		UseShellExecute = false,
		RedirectStandardOutput = true,
	};
```
... in order to trace/log out the actual values for 'executablePath', 'ownerPid', 'devToolsHost'.

This would allow me to custom code a test app using exact same parameters trying to reply the scenario + analyzing what's causing the failure.

Would this be a viable way to understand the source of trouble?

**Reply by jeromelaban** [reply]
Can you provide the information I asked for first ? 

Logging in this case will not do anything because the error raised is about the external process not providing output.

**Reply by DierkDroth** [reply]
@jeromelaban sure. Please see below

> about the external process not providing output

Understand. But is the UNO code listening only to STDOUT or to STDERR as well?

![image](https://user-images.githubusercontent.com/26179242/154853712-c83c85e5-a6ac-484c-9718-dff0d27f2549.png)

**Reply by jeromelaban** [reply]
Thanks for the update. Can you try installing .NET 5 as well and see if this helps?

**Reply by jeromelaban** [reply]
As for stderr, indeed it's not. This particular piece is a clone of the .NET original which is not listening to stderr as well, but it's worth adding it.

**Reply by DierkDroth** [reply]
@jeromelaban oh my goodness. That was it! Problem solved.

Thanks for your support

![image](https://user-images.githubusercontent.com/26179242/154989609-4025c992-797c-4766-b571-426fefc2b6bd.png)

**Reply by jeromelaban** [reply]
Thanks! We'll be pushing a bootstrapper update to avoid requiring the .NET 5 installation, as well as the stderr logging for future troubleshooting.
